### PR TITLE
Addresses Issue #4: Unit tests added for fetchData method

### PR DESCRIPTION
**Pull request by ypahalajani**
1. Made App modular (still in progress)
2. Wrote a pure function for fetching data so that it can be easily tested

### CODE REVIEW

**Comment by ypahalajani** [comment]
Address #4 
@johnsylvain , please review and let me know if I am going in the right direction. Also, I'll add a readme for running test cases. Currently, you just need to run **SpecRunner.html** (in **tests** folder) in browser. You'll see the Test Suite Status in browser.
I'll add more test cases.

**Comment by johnsylvain** [comment]
Has this been tested in the browser to ensure there are no breaking issues?

**Comment by ypahalajani** [comment]
Yes, It's working. And, these test cases are only for the fetchData method. I'll wind up this work by the end of today and let you know so that you can review the changes and merge this PR. Also, I'll write add a README for testing in this PR so that it becomes easy for others who'll work on it in near future.

PS: After you test the changes, let me know - what you think about the test cases. If everything is fine, I'll add more test cases.

**Comment by ypahalajani** [comment]
@johnsylvain: this PR is ready to be reviewed and merged!

File: app.module.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    vm.view.innerHTML = vm.renderView('main_tmpl', vm.thought)
+  }
+
+  if ((cachedThoughts &&
+    new Date() <= new Date(vm.state.cachedPosts.expirationCacheTime)) ||
+    !navigator.onLine
+  ) {
+    getAndRender(vm.state.cachedPosts.data)
+  } else {
+    this.fetchData('https://www.reddit.com/r/showerthoughts/hot.json?limit=300')
```
Comment (johnsylvain) [inline app.module.js]: Use `vm` instead of `this` for consistency

File: tab.html [hunk 2]
```diff
@@ -21,6 +21,7 @@
       <p>&#8212; <%= author %></p>
     </div>
   </script>
+  <script src="app.module.js"></script>
   <script src="background.js"></script>
```
Comment (johnsylvain) [inline tab.html]: `background.js` only has two lines of code, let's just inline the JavaScript. Example:
```html
<script>
  var app = new App();
  app.init();
</script>
```

File: app.module.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+      } else {
+        resolve(undefined)
+      }
+    });
+  });
+};
+
+App.prototype.saveState = function (state) {
+  try {
+    const serializedState = JSON.stringify(state);
```
Comment (johnsylvain) [inline app.module.js]: Use `var` instead of `const`. I want to eventually move the entire project to es6, but for now let's stay with es5 for consistency.

**Review by johnsylvain — CHANGES_REQUESTED** [review]

**Comment by johnsylvain** [comment]
Just some minor tweaks, everything else looks good.

**Comment by ypahalajani** [comment]
I've made the requested changes except for the one that said to include the two lines of app module initialization as inline script of tab.html because,

From [Chrome Extension CSP Docs](https://developer.chrome.com/extensions/contentSecurityPolicy#JSExecution), 
> Inline JavaScript will not be executed. This restriction bans both inline <script> blocks and inline event handlers (e.g. <button onclick="...">).

So, I think let's just keep it that way due to restriction. All the other changes have been updated in the PR.

BTW, I am loving this extension, Every new tab has a new quote and that too interesting ones! :+1:

File: app.module.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+  if ((cachedThoughts &&
+    new Date() <= new Date(vm.state.cachedPosts.expirationCacheTime)) ||
+    !navigator.onLine
+  ) {
+    getAndRender(vm.state.cachedPosts.data)
+  } else {
+    vm.fetchData('https://www.reddit.com/r/showerthoughts/hot.json?limit=300')
+    .then(function (res) {
+      Object.assign(vm.state.cachedPosts, {
```
Comment (johnsylvain) [inline app.module.js]: `vm.state.cache` is initially set to `undefined`, therefore `Object.assign` can throw an error in some cases (when `localStorage` is cleared). Please revert this to the original method:

```js
vm.state.cachedPosts = {
  data: res,
  expirationCacheTime: new Date().addHours(1)
}
```

**Comment by ypahalajani** [comment]
@johnsylvain: Indeed, you are right. Reverted the *Object.assign* change.

**Review by johnsylvain — APPROVED** [review]

**Pull request merged**